cloudflare/resource_cloudflare_zone.go

```diff
@@ -47,6 +47,13 @@ func resourceCloudflareZone() *schema.Resource {
 					Type: schema.TypeString,
 				},
 			},
+			"plan": {
+				Type:     schema.TypeString,
+				Computed: true,
+				Optional: true,
+				//TODO ask for complete Enum
+				ValidateFunc: validation.StringMatch(regexp.MustCompile("^([a-zA-Z0-9 ]+)$"), "Not a valid plan: use Free Website, Pro Plan, Business, Enterprise"),
```

Comment:
We should use the available values from the API for this value validation.

```suggestion
				ValidateFunc: validation.StringInSlice([]string{"free", "business", "pro", "enterprise"}, false),
```

Reply:
From what I can see in the `cloudflare-go` impl, these are `legacy_id`s, not available in the `ZoneRatePlan` returned by `client.ZoneDetails(zoneID)`. Should I add code in `resourceCloudflareZoneRead` to fetch the legacy ids from another part of the client?

Reply:
IIRC, this is a discussion we've had with our TAM and the "legacy_id" reference is named such as Cloudflare transitioned to UUIDs instead of generic strings for identifiers. 

Having a look at `AvailableRatePlans` and what `ZoneDetails` returns, there isn't a common field between the two which will make association a touch harder.

```go
// AvailableRatePlans
(cloudflare.ZoneRatePlan) {
  ID: (string) (len=10) "enterprise",
  Name: (string) (len=6) "CF_ENT",
  Price: (int) 0,
  Currency: (string) (len=3) "USD",
  Duration: (int) 0,
  Frequency: (string) (len=7) "monthly",
  Components: ([]cloudflare.zoneRatePlanComponents) (len=4 cap=4) {
   (cloudflare.zoneRatePlanComponents) {
    Name: (string) (len=22) "dedicated_certificates",
    Default: (int) 0,
    UnitPrice: (int) 5
   },
   (cloudflare.zoneRatePlanComponents) {
    Name: (string) (len=29) "dedicated_certificates_custom",
    Default: (int) 0,
    UnitPrice: (int) 10
   },
   (cloudflare.zoneRatePlanComponents) {
    Name: (string) (len=10) "page_rules",
    Default: (int) 125,
    UnitPrice: (int) 0
   },
   (cloudflare.zoneRatePlanComponents) {
    Name: (string) (len=5) "zones",
    Default: (int) 1,
    UnitPrice: (int) 0
   }
  }
 }
```

```go
// ZoneDetails
Plan: (cloudflare.ZoneRatePlan) {
  ID: (string) (len=32) "94f3b7b768b0458b56d2cac4fe5ec0f9",
  Name: (string) (len=18) "Enterprise Website",
  Price: (int) 0,
  Currency: (string) (len=3) "USD",
  Duration: (int) 0,
  Frequency: (string) (len=7) "monthly",
  Components: ([]cloudflare.zoneRatePlanComponents) <nil>
 }
```

Perhaps instead of trying to associate the two through existing methods, we just statically define the plan levels (free, enterprise, etc) and then have a fuzzy match function that ties it to an existing plan and errors if it doesn't match? Similar to your `setRatePlan` method but just using a static definition of available plans. Here is a diff demonstrating the idea

```diff
diff --git a/cloudflare/resource_cloudflare_zone.go b/cloudflare/resource_cloudflare_zone.go
index eeea482..51b308e 100644
--- a/cloudflare/resource_cloudflare_zone.go
+++ b/cloudflare/resource_cloudflare_zone.go
@@ -10,8 +10,11 @@ import (
 	cloudflare "github.com/cloudflare/cloudflare-go"
 	"github.com/hashicorp/terraform/helper/schema"
 	"github.com/hashicorp/terraform/helper/validation"
+	"github.com/pkg/errors"
 )
 
+var validPlanTypes = []string{"free", "business", "pro", "enterprise"}
+
 func resourceCloudflareZone() *schema.Resource {
 	return &schema.Resource{
 		Create: resourceCloudflareZoneCreate,
@@ -51,8 +54,7 @@ func resourceCloudflareZone() *schema.Resource {
 				Type:         schema.TypeString,
 				Computed:     true,
 				Optional:     true,
-				//TODO ask for complete Enum
-				ValidateFunc: validation.StringMatch(regexp.MustCompile("^([a-zA-Z0-9 ]+)$"), "Not a valid plan: use Free Website, Pro Plan, Business, Enterprise"),
+				ValidateFunc: validation.StringInSlice(validPlanTypes, false),
 			},
 			"meta": {
 				Type:     schema.TypeMap,
@@ -115,10 +117,6 @@ func resourceCloudflareZoneCreate(d *schema.ResourceData, meta interface{}) erro
 		}
 	}
 
-	if err := setRatePlan(d, client, zone.ID); err != nil {
-		return err
-	}
-
 	return resourceCloudflareZoneRead(d, meta)
 }
 
@@ -147,7 +145,12 @@ func resourceCloudflareZoneRead(d *schema.ResourceData, meta interface{}) error
 	d.Set("name_servers", zone.NameServers)
 	d.Set("meta", flattenMeta(d, zone.Meta))
 	d.Set("zone", zone.Name)
-	d.Set("plan", zone.Plan.Name)
+
+	planName, err := normalisePlanName(zone.Plan.Name)
+	if err != nil {
+		return err
+	}
+	d.Set("plan", planName)
 
 	return nil
 }
@@ -168,10 +171,6 @@ func resourceCloudflareZoneUpdate(d *schema.ResourceData, meta interface{}) erro
 		}
 	}
 
-	if err := setRatePlan(d, client, zoneID); err != nil {
-		return err
-	}
-
 	return resourceCloudflareZoneRead(d, meta)
 }
 
@@ -201,28 +200,22 @@ func flattenMeta(d *schema.ResourceData, meta cloudflare.ZoneMeta) map[string]in
 	return cfg
 }
 
-func setRatePlan(d *schema.ResourceData, client *cloudflare.API, zoneID string) error {
-	if planName, ok := d.GetOk("planName"); ok {
-		ratePlan, err := getZonePlanIDFor(client, zoneID, planName.(string))
-		if err != nil {
-			return fmt.Errorf("Error fetching planName %s for zone %q: %s", planName, zoneID, err)
-		}
-		if _, err := client.ZoneSetRatePlan(zoneID, *ratePlan); err != nil {
-			return fmt.Errorf("Error setting planName %s for zone %q: %s", planName, zoneID, err)
-		}
-	}
-	return nil
+func normalisePlanName(planName string) (string, error) {
+	planName = strings.ToLower(planName)
+	planName = strings.Replace(planName, " website", "", 1)
+
+	if !sliceContains(planName, validPlanTypes) {
+		return "", errors.New("invalid plan name detected")
 	}
 
-func getZonePlanIDFor(client *cloudflare.API, zoneID, planName string) (*cloudflare.ZoneRatePlan, error) {
-	plans, err := client.AvailableZoneRatePlans(zoneID)
-	if err != nil {
-		return nil, err
+	return planName, nil
 }
-	for _, p := range plans {
-		if strings.EqualFold(p.Name, planName) {
-			return &p, nil
+
+func sliceContains(a string, list []string) bool {
+	for _, b := range list {
+		if b == a {
+			return true
 		}
 	}
-	return nil, fmt.Errorf("Not found amongst plans")
+	return false
 }
diff --git a/cloudflare/resource_cloudflare_zone_test.go b/cloudflare/resource_cloudflare_zone_test.go
index 45043e6..ff01392 100644
--- a/cloudflare/resource_cloudflare_zone_test.go
+++ b/cloudflare/resource_cloudflare_zone_test.go
@@ -20,7 +20,7 @@ func TestZone(t *testing.T) {
 					resource.TestCheckResourceAttr(name, "zone", "example.org"),
 					resource.TestCheckResourceAttr(name, "paused", "true"),
 					resource.TestCheckResourceAttr(name, "name_servers.#", "2"),
-					resource.TestCheckResourceAttr(name, "plan", "Free Website"),
+					resource.TestCheckResourceAttr(name, "plan", "free"),
 				),
 			},
 		},

```

While looking into this, I've found that `resourceCloudflareZoneUpdate` doesn't actually do much of an update right now (just pauses the domain). I'll open a separate issue for this.

Reply:
Looks good, just when we are creating the zone in `resourceCloudflareZoneCreate` how do we instruct the zone to be assigned the desired plan?

Reply:
No need for fuzzy logic. You may use `AvailableZoneRatePlans` to get the list of Plans that can be applied to a zone, including their IDs, which can be used to `PATCH` the zone resource.

Reply:
You cannot do it in single step, you'll need to edit the zone after creation with `ZoneSetRatePlan` or `EditZone`.

Reply:
Agreed :) I will leave the `setRatePlan` function in the create and update function